protocol/igmp.go

```diff
@@ -0,0 +1,421 @@
+		MaxResponseTime: maxResponseTime,
+		GroupAddress:    group,
+		IntervalTime:    queryInterval,
+		NumberOfSources: uint16(len(sources)),
+		SourceAddresses: sources,
+	}
+}
+
+// IGMPv3GroupRecord:
+//    0                   1                   2                   3
+//    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |  Record Type  |  Aux Data Len |     Number of Sources (N)     |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                       Multicast Address                       |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                       Source Address [1]                      |
+//   +-                                                             -+
+//   |                       Source Address [2]                      |
+//   +-                                                             -+
+//   .                               .                               .
+//   .                               .                               .
+//   .                               .                               .
+//   +-                                                             -+
+//   |                       Source Address [N]                      |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//   |                                                               |
+//   .                                                               .
+//   .                         Auxiliary Data                        .
+//   .                                                               .
+//   |                                                               |
+//   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type IGMPv3GroupRecord struct {
+	Type             uint8
+	AuxDataLen       uint8 // this should always be 0 as per IGMPv3 spec.
+	NumberOfSources  uint16
+	MulticastAddress net.IP
+	SourceAddresses  []net.IP
+	AuxData          []uint32 // NOT USED
+}
+
+func (p *IGMPv3GroupRecord) Len() uint16 {
+	return 8 + uint16(p.AuxDataLen)*4 + p.NumberOfSources*4
+}
+
+func (p *IGMPv3GroupRecord) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, int(p.Len()))
+	n := 0
+	data[n] = p.Type
+	n += 1
+	data[n] = p.AuxDataLen
+	n += 1
+	binary.BigEndian.PutUint16(data[n:], p.NumberOfSources)
+	n += 2
+	copy(data[n:n+4], p.MulticastAddress.To4())
+	n += 4
+	for _, d := range p.AuxData {
+		binary.BigEndian.PutUint32(data[n:], d)
+		n += 4
+	}
+	return
+}
+
+func (p *IGMPv3GroupRecord) UnmarshalBinary(data []byte) error {
+	if len(data) < 8 {
+		return errors.New("The []byte is too short to unmarshal a full IGMPv3GroupRecord message.")
+	}
+	n := 0
+	p.Type = data[n]
+	n += 1
+	p.AuxDataLen = data[n]
+	n += 1
+	p.NumberOfSources = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	p.MulticastAddress = data[n : n+4]
+	n += 4
```

Comment:
Please add the condition to check for end of buffer as you did for IGMPv3Query's UnmarshalBinary.